### PR TITLE
Store expression hierarchy in semantic model snapshots

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

When we iterate over the AST for analysis, we often process nodes in a "deferred" manner. For example, if we're analyzing a function, we push the function body onto a deferred stack, along with a snapshot of the current semantic model state. Later, when we analyze the body, we restore the semantic model state from the snapshot. This ensures that we know the correct scope, hierarchy of statement parents, etc., when we go to analyze the function body.

Historically, we _haven't_ included the _expression_ hierarchy in the model snapshot -- so we track the current expression parents in the visitor, but we never save and restore them when processing deferred nodes. This can lead to subtle bugs, in that methods like `expr_parent()` aren't guaranteed to be correct, if you're in a deferred visitor.

This PR migrates expression tracking to mirror statement tracking exactly. So we push all expressions onto an `IndexVec`, and include the current expression on the snapshot. This ensures that `expr_parent()` and related methods are "always correct" rather than "sometimes correct".

There's a performance cost here, both at runtime and in terms of memory consumption (we now store an additional pointer for every expression). In my hyperfine testing, it's about a 1% performance decrease for all-rules on CPython (up to 533.8ms, from 528.3ms) and a 4% performance decrease for default-rules on CPython (up to 212ms, from 204ms). However... I think this is worth it given the incorrectness of our current approach. In the future, we may want to reconsider how we do these upward traversals (e.g., with something like a red-green tree). (**Note**: in https://github.com/astral-sh/ruff/pull/6351, the slowdown seems to be entirely removed.)

### CODE REVIEW

**Comment by charliermarsh** [comment]
This will also fix https://github.com/astral-sh/ruff/issues/6285, but I'll handle that in a separate PR, I want to update the fixtures and it would be too noisy here.

**Comment by charliermarsh** [comment]
Hmm, we don't need back references for expressions (we never need to be able to go from a node to its ID, or from a node to its parent -- we can always go from an ID to its parent), so perhaps we can improve performance here by omitting the hash map from `Nodes` in that case. I will test it out.

**Comment by charliermarsh** [comment]
Ok, I'm going to resolve the performance degradation in a separate PR because I think it's easier to review this as-is.

File: crates/ruff/src/rules/flake8_pytest_style/rules/assertion.rs [hunk 1]
```diff
@@ -248,7 +248,6 @@ pub(crate) fn unittest_assertion(
                     // the assertion is part of a larger expression.
                     if checker.semantic().stmt().is_expr_stmt()
                         && checker.semantic().expr_parent().is_none()
-                        && !checker.semantic().scope().kind.is_lambda()
```
Comment (MichaReiser) [inline crates/ruff/src/rules/flake8_pytest_style/rules/assertion.rs:251]: Is this change intentional?

File: crates/ruff_python_semantic/src/model.rs [hunk 2]
```diff
@@ -36,8 +36,11 @@ pub struct SemanticModel<'a> {
     /// The identifier of the current statement.
     stmt_id: Option<NodeId>,
 
-    /// Stack of current expressions.
-    exprs: Vec<&'a Expr>,
+    /// Stack of all visited expressions.
+    exprs: Nodes<'a, Expr>,
+
+    /// The identifier of the current expression.
+    expr_id: Option<NodeId>,
```
Comment (MichaReiser) [inline crates/ruff_python_semantic/src/model.rs:43]: Nit: We now store both a stack of statements and expressions. This gives us almost a full fidelity parent chain, except that it isn't possible to get the `parent` statement of an expression (except for the current expression). 

What do you think of introducing a `Vec<AnyNodeRef>` that stores all parent nodes (Ultimately means `O(n)` pointers to be written, where `n` = len(AST)) instead? You can retrieve the current statement by traversing upwards until you find the first `is_statement()` node. 

The one issue of that is that there's currently no way to go from `AnyNodeRef` to `&Stmt`. We would need to introduce a `AnyStatementRef` enum for that.

File: crates/ruff_python_semantic/src/model.rs [hunk 3]
```diff
@@ -836,22 +839,32 @@ impl<'a> SemanticModel<'a> {
...
-    pub fn expr_ancestors(&self) -> impl Iterator<Item = &&Expr> {
-        self.exprs.iter().rev().skip(1)
+    pub fn expr_ancestors(&self) -> impl Iterator<Item = &'a Expr> + '_ {
+        self.expr_id
+            .iter()
+            .copied()
+            .flat_map(|id| {
+                self.exprs
+                    .ancestor_ids(id)
+                    .skip(1)
```
Comment (MichaReiser) [inline crates/ruff_python_semantic/src/model.rs:864]: Why is it necessary to `skip(1)`? Shouldn't `ancestor_ids` return the ancestors only?

File: crates/ruff_python_semantic/src/model.rs [hunk 4]
```diff
@@ -836,22 +839,32 @@ impl<'a> SemanticModel<'a> {
...
+                    .map(|id| &self.exprs[id])
+            })
+            .copied()
```
Comment (MichaReiser) [inline crates/ruff_python_semantic/src/model.rs:867]: Nit: Use successors? 

```suggestion
std::iter::successors(self.expr_id, |id| self.exprs.ancestor_ids(id).skip(1).map(|id| self.exprs[id])))
```

**Review by MichaReiser — APPROVED** [review]

**Comment by MichaReiser** [comment]
Could we add some tests for the new introduced APIs? Or how did you test that this works as expected?

File: crates/ruff_python_semantic/src/model.rs (hunk 2 above)
Comment (charliermarsh) [reply crates/ruff_python_semantic/src/model.rs:43]: I will consider this separately, it makes sense to me, although it may come with the performance regression that I documented here and fixed in https://github.com/astral-sh/ruff/pull/6351.

File: crates/ruff/src/rules/flake8_pytest_style/rules/assertion.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/ruff/src/rules/flake8_pytest_style/rules/assertion.rs:251]: Yes! Not necessary. See the identical change in `crates/ruff/src/rules/pandas_vet/rules/inplace_argument.rs` that removes the large comment around Case 3. The fact that these tests still pass is a evidence that this change is working as expected.

File: crates/ruff_python_semantic/src/model.rs (hunk 3 above)
Comment (charliermarsh) [reply crates/ruff_python_semantic/src/model.rs:864]: Right now, `ancestor_ids` starts with the current ID IIRC, probably because it's convenient with the `successors` API.  Do you find that confusing?

**Comment by charliermarsh** [comment]
> Could we add some tests for the new introduced APIs? Or how did you test that this works as expected?

I'm going to add some tests in a separate PR, I wanted to make sure that this had a clean snapshot suite.

File: crates/ruff_python_semantic/src/model.rs (hunk 4 above)
Comment (charliermarsh) [reply crates/ruff_python_semantic/src/model.rs:867]: This doesn't quite do the same thing. The awkwardness is that `self.expr_id` can be `None`.

File: crates/ruff_python_semantic/src/model.rs (hunk 3 above)
Comment (charliermarsh) [reply crates/ruff_python_semantic/src/model.rs:864]: (This is current behavior on `main` so I'll look into it separately.)

File: crates/ruff_python_semantic/src/model.rs (hunk 4 above)
Comment (MichaReiser) [reply crates/ruff_python_semantic/src/model.rs:867]: I'm not sure if I understand. The first argument to `std::iter::successors` is an `Option`. Both `successors and `expr_id.flat_map` should short circuit if expr_id is `None`.

Comment (charliermarsh) [reply crates/ruff_python_semantic/src/model.rs:867]: Oh, I think you just had misplaced parentheses in your suggestion and I misunderstood.

File: crates/ruff_python_semantic/src/model.rs (hunk 2 above)
Comment (charliermarsh) [reply crates/ruff_python_semantic/src/model.rs:43]: I was looking into this and I want to understand the last paragraph here. Is the suggestion that methods like `SemanticModel#current_statement` would now return `AnyStatementRef`? And so we'd change all consumers of that method to work with `AnyStatementRef` rather than `&Stmt`?

Comment (charliermarsh) [reply crates/ruff_python_semantic/src/model.rs:43]: If so, we also need `AnyExpressionRef`, and I'd ultimately like to store the full-fidelity chain here including `Parameters`, `Arguments`, etc...

Comment (charliermarsh) [reply crates/ruff_python_semantic/src/model.rs:43]: E.g., these kinds of changes: https://github.com/astral-sh/ruff/compare/charlie/any-node-ref?expand=1#diff-c2c8422959405016e1fb54ca23edee9716bd06ad64e610ceea9612a8291122fcR1815. It's feasible but a lot of call sites to work through, so I want to make sure I understand what's being suggested before I think on it further.

Comment (charliermarsh) [reply crates/ruff_python_semantic/src/model.rs:43]: I might propose that we do this incrementally by adding an internal type to the semantic crate that's like `AnyNodeRef`, but less granular:

```rust
#[derive(Debug, Clone)]
pub enum AnyNodeRef<'a> {
    Stmt(&'a Stmt),
    Expr(&'a Expr),
}

impl<'a> AnyNodeRef<'a> {
    pub fn as_statement(&self) -> Option<&'a Stmt> {
        match self {
            AnyNodeRef::Stmt(stmt) => Some(stmt),
            AnyNodeRef::Expr(_) => None,
        }
    }
}
```

That way, we wouldn't need to change the `SemanticModel` interface at all for now.
